lib/rules/arrow-parens.js

```diff
@@ -80,29 +80,49 @@ module.exports = {
          * @returns {void}
          */
         function parens(node) {
-            const isAsync = node.async;
-            const firstTokenOfParam = sourceCode.getFirstToken(node, isAsync ? 1 : 0);
+            const firstToken = sourceCode.getFirstToken(node);
+            const arrowToken = sourceCode.getTokenBefore(node.body, astUtils.isArrowToken);
+            const openingParenToken = astUtils.isOpeningParenToken(firstToken)
+                ? firstToken
+                : sourceCode.getFirstTokenBetween(firstToken, arrowToken, { filter: astUtils.isOpeningParenToken });
```

Comment:
Can this find the wrong opening paren token, for example in a function such as this:

```js
<T extends (A|B)>() => { }
```

Reply:
I've added additional test case for this scenario, seems to be working fine

Reply:
I think the added `"<T extends (A | B)>() => { }"`  test case isn't correct as `generic-no-params` 'parser' returns AST and tokens for another code - `"<T extends Object>(): T => { return 1 }"`?

But that's okay, let's not add new tests for now. Maybe there is no need, or maybe there could be a better example than this one.

The example was just for a discussion - can `openingParenToken` be something that actually isn't the paren for parameters, could something go wrong in that case, and would it be maybe more clear to ensure that `openingParenToken` is indeed the paren for parameters (if it exists).

Reply:
Oh yes, totally forgot about mock AST tree for that case. Should I just revert it for now and `openingParenToken` thingy should get another issue opened? 

I'm not even sure if this is possible case though, since you cannot use parenthesis with union types (that's not an OR statement where the order matters)

Reply:
> Oh yes, totally forgot about mock AST tree for that case. Should I just revert it for now and `openingParenToken` thingy should get another issue opened?

I agree, let's revert that change for now, i.e. remove the `"<T extends (A | B)>() => { }"` test.

> I'm not even sure if this is possible case though, since you cannot use parenthesis with union types (that's not an OR statement where the order matters)

I think it's a valid syntax to have parentheses there, and they can have a meaning in some cases, e.g., `(A | B) & C` is not the same as `A | B & C`?

Reply:
Like I said, I believe union operator doesn't have priority on it, so parenthesis don't make sense, you can try here:

```
type A = {
    a: string;
}

type B = {
    b: string;
}

type C = {
    c: string;
}

type T = A | B & C

const test: T = {
    a: 'a',
    b: 'b',
    c: 'c'
}
```

[Playground Link](https://www.typescriptlang.org/play/?emitDecoratorMetadata=true#code/C4TwDgpgBAglC8UDeAoK6oEMBcUDOwATgJYB2A5gNwoC+KKokUAQgsmhgEa4EkXV0G4aAGE2qDFADGPImSq16jaABU2cAD4soAMigj6UgPakCUYBAK41iCRhxQA5JkcAaDum5PObj9NyOUo60QA)

You can see that modifying the T type with parenthesis doesn't make any difference in the T type. It's not a logical OR/AND, where order makes a difference as well as grouping. 

I will revert the test once I'm back home 👍

Reply:
I still think there is a difference:

```js
type T = A | B & C

// valid, it's A
const test: T = {
    a: 'a'
}
```

vs

```js
type T = (A | B) & C

// invalid, it isn't A & C
const test: T = {
    a: 'a'
}
```

Reply:
yup, my bad, I got it breaking now 👍 So, I will revert the test case for this, and I guess create new issue to cover this behavior in another PR?

Reply:
There is something else, it looks like the change in this PR can introduce some new problems.

For example, the following code has no eslint warnings in the actual master version (by chance):

```js
/* eslint arrow-parens: ["error", "as-needed"] */

<T>(x) => {};
```

after this change, parens around `x` will be seen as unnecessary, and the code will be auto-fixed to:

```js
/* eslint arrow-parens: ["error", "as-needed"] */

<T>x => {};
```

which is invalid syntax in TS, parens were indeed necessary.

It will also remove all parens in `<>` as well, e.g., this is no error at the moment:

```js
/* eslint arrow-parens: ["error", "as-needed"] */

<T extends (A | B) & C>(x) => {};
```

but after the change it will be auto-fixed twice, and the first iteration will remove B:

```js
/* eslint arrow-parens: ["error", "as-needed"] */

<T extends A & C>x => {};
```

Would it be possible to avoid these problems? It probably isn't common to have a generic function without type annotations for param and/or return value, but it would nice to avoid introducing new bugs while fixing the existing ones.

Reply:
Sure, I will add test cases for typescript with new AST tree for those cases and fix the issue

Reply:
@mdjermanovic even though the first example is fine in the TS playground, I cannot get https://astexplorer.net/ to make it work with @typescript/eslint-parser to generate me AST tree and tokens for the mock data: `JSX element 'T' has no corresponding closing tag.` Any clues what I might be doing wrong?

Reply:
AST explorer enables some parser options by default. It seems that `@typescript-eslint/parser` treats `<T>` without `extends` as an opening JSX tag when JSX parsing is enabled.

There is a settings icon next to the parser's name where you can customize parser options and turn off `jsx` in the `ecmaFeatures` section:

![image](https://user-images.githubusercontent.com/44349756/71603746-ae55cc00-2b5e-11ea-89ab-210f81c1b0ef.png)

Reply:
So, I've added 3 more test cases for typescript and code fixes to address regression introduced by this refactor, as asked. For now I just put the `isTokenPartOfNode` helper inline in the linting rule, if you think the overall idea is fine on how I find the openingPerenToken, let me know and I will move out the `isTokenPartOfNode` to `astUtils` and add some unit tests for it as well.

Reply:
I left a couple of notes about what might need to be changed in the approach on how to find `openingParenToken`. 

Nevertheless, `isTokenPartOfNode` seems useful and probably even necessary for this :+1: